src/test/mds/TestQuiesceDb.cc

```diff
@@ -0,0 +1,824 @@
+        r.expiration = sec(100);
+      }
+      r.include_roots({"root1"});
+      r.await = sec(100);
+    }));
+  }
+
+  // flush the pending requests by running a simple query
+  ASSERT_EQ(OK(), run_request([](auto&r){r.query("set1");}));
+
+  ASSERT_EQ(awaits.size(), managers.at(0)->internal_awaits().size());
+
+  for (auto&& await: awaits) {
+    EXPECT_EQ(NA(), await->check_result()) << await->request.set_id.value();
+  }
+
+  QuiesceDbVersion quiescing_v;
+  // wait for the replicas to ack the quiescing root
+  {
+    std::unique_lock l(comms_mutex);
+
+    // let two of the three peers ack quiescing of the root
+    managers.at(0)->reset_agent_callback(QUIESCING_AGENT_CB);
+    managers.at(1)->reset_agent_callback(QUIESCING_AGENT_CB);
+
+    auto result = comms_cond.wait_for(l, std::chrono::milliseconds(10000), [&] {
+      auto& ack0 = latest_acks[0];
+      auto& ack1 = latest_acks[1];
+      quiescing_v = std::max(ack0.version, ack1.version);
+      return ack0.roots.size() > 0 && ack1.roots.size() > 0;
+    });
+    ASSERT_TRUE(result);
+  }
+
+  // kick the db queue with a simple query
+  ASSERT_EQ(OK(), run_request([](auto& r) { r.query("set1"); }));
+  EXPECT_EQ(QS_QUIESCING, last_request->response.sets.at("set1").rstate.state);
+
+  // should still be waiting for the last agent
+  for (auto&& await: awaits) {
+    EXPECT_EQ(NA(), await->check_result()) << await->request.set_id.value();
+  }
+
+  managers.at(2)->reset_agent_callback(QUIESCING_AGENT_CB);
+
+  // wait for the last manager to ack
+  {
+    std::unique_lock l(comms_mutex);
+    auto result = comms_cond.wait_for(l, std::chrono::milliseconds(1000), [&] {
+      auto& ack2 = latest_acks[2];
+      return ack2.version > quiescing_v;
+    });
+    ASSERT_TRUE(result);
+  }
+
+  // kick the db queue with a simple query
+  ASSERT_EQ(OK(), run_request([](auto& r) {}));
+
+  // first three sets must be expired because they had 0 expiration
+  EXPECT_EQ(QS_EXPIRED, last_request->response.sets.at("set1").rstate.state);
+  EXPECT_EQ(QS_EXPIRED, last_request->response.sets.at("set2").rstate.state);
+  EXPECT_EQ(QS_EXPIRED, last_request->response.sets.at("set3").rstate.state);
+  // set4 was configured with non-zero expiration, hence we expect it to be quiesced
+  EXPECT_EQ(QS_QUIESCED, last_request->response.sets.at("set4").rstate.state);
+
+  // pending quiesce requests must have all completed successfully
+  // even though some of the sets got expired immediately
+  for (auto&& await: awaits) {
+    EXPECT_EQ(OK(), await->check_result()) << await->request.set_id.value();
+  }
+
+  // prevent one of the acks
+  managers.at(1)->reset_agent_callback();
+
+  ASSERT_EQ(OK(), run_request([](auto& r) {
+    r.set_id = "set4";
+    r.release_roots();
+  }));
+
+  // shouldn't hurt to run twice
+  ASSERT_EQ(OK(), run_request([](auto& r) {
+    r.set_id = "set4";
+    r.release_roots();
+  }));
+
+  EXPECT_EQ(QS_RELEASING, last_request->response.sets.at("set4").rstate.state);
```

Comment:
multi-rank releasing can be separate test

Reply:
ack